app/models/workflow_step.rb

```diff
@@ -34,16 +34,25 @@ def as_milestone(xml)
                   version: version)
   end
 
-  # callback to set the completed_at column to the current time if we are completing a step
-  def set_completed_at
+  # callback to set the completed_at column to the current time if we are
+  # completing a step and completed_at is not already set
+  def maybe_set_completed
+    return unless completed?
+    return if already_completed?
+
     self.completed_at = Time.now
   end
 
   ##
   # indicate if this step is marked as completed
   # @return [boolean]
   def completed?
-    COMPLETED_STATES.include? status
+    COMPLETED_STATES.include?(status)
+  end
+
+  # Check if there is already a completed_at date
+  def already_completed?
```

Comment:
I think the fact that there is a `completed?` method and an `already_completed?` method is confusing.

Reply:
agree, change made